### PR TITLE
Refaсtor operator controllers

### LINKED ISSUE #115: Refactor Operator Controllers
ATM we have 3 Controller in the Operator `ServiceClusterRegistration`, `Ferry` and `Catapult` and all controllers are copies of the same code.

When someone is bored, we should refactor these controllers to have less code duplication.

### PR DESCRIPTION
**Pull request by aborilov**
**What this PR does / why we need it**:

Added `BaseController` that provides common controller logic. It
provides methods:
 * Reconcile
 * FetchObject
 * HandleDeletion
 * UpdateStatus
 * ReconcileOwnedObjects

`Reconcile` method provides common reconcile logic, you can call it or
use other methods to compile own reconcile logic.

`BaseController` methods take as parameter `Controller` interface with
methods

```
GetObj() object
GetManifests(context.Context) ([]unstructured.Unstructured, error)
GetReadyConditionStatus() operatorv1alpha1.ConditionStatus
SetTerminatingCondition(context.Context) bool
SetUnReadyCondition()
SetReadyCondition()
```

To add a new operator controller we just need to implement those methods.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #115

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/operator/internal/controllers/kubecarrier.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+type KubeCarrierStrategy struct {
+}
+
+func (c *KubeCarrierStrategy) GetObj() Component {
+	return &operatorv1alpha1.KubeCarrier{}
+}
+
+func (c *KubeCarrierStrategy) GetOwnedObjectsTypes() []runtime.Object {
+	return []runtime.Object{
```
[the remark this answers is not in the document]
Comment (aborilov) [reply pkg/operator/internal/controllers/kubecarrier.go]: @nmiculinic @jiachengxu should I rename `GetOwnedObjectsTypes ` to `GetDeletionObjectsTypes`?

[the remark this answers is not in the document]
Comment (nmiculinic) [reply pkg/operator/internal/controllers/kubecarrier.go]: these are the objects you need to clear before removing the finalizer? let's call it then `GetDeletionBlockingObjectTypes` and be done with it?

[the remark this answers is not in the document]
Comment (nmiculinic) [reply pkg/operator/internal/controllers/kubecarrier.go]: or `GetDeletionObjectTypes` is fine as well

**Comment by nmiculinic** [comment]
This looks good from my POV; @jiachengxu  can we merge it?

File: pkg/operator/internal/controllers/catapult.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func (c *CatapultStrategy) AddWatches(builder *builder.Builder, scheme *runtime.Scheme) *builder.Builder {
+	return builder.
+		Owns(&appsv1.Deployment{}).
+		Owns(&corev1.Service{}).
+		Owns(&rbacv1.Role{}).
+		Owns(&rbacv1.RoleBinding{}).
+		Owns(&certv1alpha2.Issuer{}).
+		Owns(&certv1alpha2.Certificate{})
```
Comment (jiachengxu) [inline pkg/operator/internal/controllers/catapult.go]: Please add an `Owns` for `ServiceAccount`.

File: pkg/operator/internal/controllers/elevator.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+func (c *ElevatorStrategy) GetObj() Component {
+	return &operatorv1alpha1.Elevator{}
+}
+
+func (c *ElevatorStrategy) GetDeletionObjectTypes() []runtime.Object {
+	return []runtime.Object{
+		&rbacv1.ClusterRole{},
+		&rbacv1.ClusterRoleBinding{},
+		&corev1.ServiceAccount{},
```
Comment (jiachengxu) [inline pkg/operator/internal/controllers/elevator.go]: As I mentioned, the `ServiceAccount` is in the same namespace with the owner, so please put it in the `AddWatches`

File: pkg/operator/internal/controllers/kubecarrier.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func (c *KubeCarrierStrategy) AddWatches(builder *builder.Builder, scheme *runtime.Scheme) *builder.Builder {
+	return builder.
+		Owns(&appsv1.Deployment{}).
+		Owns(&corev1.Service{}).
+		Owns(&rbacv1.Role{}).
+		Owns(&rbacv1.RoleBinding{}).
+		Owns(&certv1alpha2.Issuer{}).
+		Owns(&certv1alpha2.Certificate{})
```
Comment (jiachengxu) [inline pkg/operator/internal/controllers/kubecarrier.go]: Please add `ServiceAccount` here

**Pull request merged**